### PR TITLE
MoE Triton Kernel

### PR DESCRIPTION
**Pull request by Mecoli1219**
## Summary
<!--- This is a required section; please describe the main purpose of this proposed code change. --->
Triton implementation of a fused MoE expert kernel for training, inspired by SonicMoE: https://arxiv.org/abs/2512.14080                                                         
                                                                                                                                                                                    
  Related to: #958

## Details

**Known limitation:** The current implementation scales well with number of experts (more parallelism) but not with number of tokens. For models with few experts (e.g., Mixtral-8x7B) at large token counts, the eager per-expert loop is likely faster.

**Convergence:** the convergence test uses wider tolerances for MoE models patched with this kernel. 

## Benchmarking Results

- Model: Qwen3-30B-A3B
- #Tokens: 8192 | #Experts: 128 | top-K: 8 | hidden: 2048 | intermediate: 768

| Sweep Dim | #Tokens | #Experts | 
| :--- | --- | --- |
| Memory | <img width="1000" height="600" alt="fused_moe_memory_full_token_length" src="https://github.com/user-attachments/assets/5c55e185-ff15-4d1a-9aab-c0c315f94ea0" /> | <img width="1000" height="600" alt="fused_moe_memory_full_token_length" src="https://github.com/user-attachments/assets/f274f84f-8a7e-4c67-900f-1b96ca5b80d0" /> |
| Speed (Full) | <img width="1000" height="600" alt="fused_moe_speed_full_token_length" src="https://github.com/user-attachments/assets/4350afdc-2820-4748-bd0c-7d05812ddc60" /> | <img width="1000" height="600" alt="fused_moe_speed_full_token_length" src="https://github.com/user-attachments/assets/d537c312-af6d-407f-9271-962b54284cd1" /> |
| Speed (Forward) | <img width="1000" height="600" alt="fused_moe_speed_forward_token_length" src="https://github.com/user-attachments/assets/8256910e-1910-4117-adcf-44cd91e6125a" /> | <img width="1000" height="600" alt="fused_moe_speed_forward_token_length" src="https://github.com/user-attachments/assets/4ece35dd-215d-4c86-a6b6-f92d60c8c22c" /> |
| Speed (Backward) | <img width="1000" height="600" alt="fused_moe_speed_backward_token_length" src="https://github.com/user-attachments/assets/2a12f652-9424-48be-bce8-b5e8c1a8289d" /> | <img width="1000" height="600" alt="fused_moe_speed_backward_token_length" src="https://github.com/user-attachments/assets/69e308bb-51e0-4ec4-bb08-e6cf3f87c0ce" /> |






### Benchmarking with patching into Qwen3-30B-A3B
- Parameters
  - Max sequence length: 32768
  - Per device batch size: 1
  - Gradient accumulate steps: 8
  - GPU: 2 * H100_8
- Result
  - Processing **8.24x** more tokens per second
  - Training speed up: 8.19x
  - Slightly save memory: -1%
  - Eval speed up: 4.1x












<!-- 
Replace BLANK with your device type. For example, A100-80G-PCIe

Complete the following tasks before sending your PR, and replace `[ ]` with
`[x]` to indicate you have done them. 
-->

- Hardware Type: H100
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by Mecoli1219** [comment]
@vaibhavjindal @Tcc0403 @kolehma8 the PR is ready to review.

File: test/transformers/test_fused_moe.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+from liger_kernel.ops.fused_moe import LigerFusedMoEFunction
+from liger_kernel.ops.fused_moe import compute_routing_metadata
+
+# ---------------------------------------------------------------------------
+# Skip if no CUDA
+# ---------------------------------------------------------------------------
+
+pytestmark = pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA required")
+
+
```
Comment (Tcc0403) [inline test/transformers/test_fused_moe.py]: There are also xpu and npu besides cuda platform.

File: test/transformers/test_fused_moe.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+)
+@pytest.mark.parametrize(
+    "dtype, atol, rtol",
+    [
+        (torch.float32, 1e-3, 1e-4),
+        (torch.bfloat16, 1e-1, 1e-2),
+    ],
+)
+def test_forward_correctness(T, E, H, intermediate_dim, K, dtype, atol, rtol):
+    device = "cuda"
```
Comment (Tcc0403) [inline test/transformers/test_fused_moe.py]: `device = infer_devce()` in global scope

File: test/transformers/test_fused_moe.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        (7, 4, 32, 16, 2),  # T < BLOCK_M_TOKEN: sub-tile padding edge case
+        (512, 8, 128, 64, 2),  # multi-tile baseline: T*K/E=128 → 2 tiles/expert
+        (512, 8, 97, 47, 2),  # multi-tile + odd H/I: tail masking in dW1, dW2, dX_expanded
+        (512, 7, 64, 32, 3),  # multi-tile + prime E: non-pow2 dW grid decomposition
+        (512, 8, 64, 32, 1),  # multi-tile + K=1: dx reduction is trivial gather
+    ],
+)
+def test_backward_correctness(T, E, H, intermediate_dim, K):
+    """Compare gradients of fused kernel vs. reference implementation."""
+    device = "cuda"
```
Comment (Tcc0403) [inline test/transformers/test_fused_moe.py]: Ditto

File: test/transformers/test_fused_moe.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+)
+def test_forward_correctness(T, E, H, intermediate_dim, K, dtype, atol, rtol):
+    device = "cuda"
+    x, gate_up_proj, down_proj, top_k_index, top_k_weights = _make_inputs(T, E, H, intermediate_dim, K, dtype, device)
+
+    ref = _reference_moe_forward(x, gate_up_proj, down_proj, top_k_index, top_k_weights)
+    out = LigerFusedMoEFunction.apply(x, gate_up_proj, down_proj, top_k_index, top_k_weights)
+
+    assert out.shape == ref.shape, f"Shape mismatch: {out.shape} vs {ref.shape}"
+    assert torch.allclose(out, ref, atol=atol, rtol=rtol), (
```
Comment (Tcc0403) [inline test/transformers/test_fused_moe.py]: https://docs.pytorch.org/docs/2.11/testing.html#torch.testing.assert_close

Use `torch.testing.assert_close` or `utils.assert_verbose_allclose` in our test suite

File: test/transformers/test_fused_moe.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    ],
+)
+@pytest.mark.parametrize(
+    "dtype, atol, rtol",
+    [
+        (torch.float32, 1e-3, 1e-4),
+        (torch.bfloat16, 1e-1, 1e-2),
+    ],
+)
+def test_forward_correctness(T, E, H, intermediate_dim, K, dtype, atol, rtol):
```
Comment (Tcc0403) [inline test/transformers/test_fused_moe.py]: Most kernel correctness tests in our test suite are forward+backward in one test. Is there any concern about separating them, or should we follow the convention for consistency?

File: test/transformers/test_swiglu.py [hunk 6]
```diff
@@ -278,27 +281,31 @@ def test_correctness_mixtralexperts(bsz, seq_len, hidden_size, intermediate_size
...
+        print(
+            f"\n  [{label}]"
+            f"\n    max_abs={diff.max():.4g}  ref={af.flatten()[abs_idx]:.4g}  liger={bf.flatten()[abs_idx]:.4g}"
+            f"\n    max_rel={rel.max():.4g}   ref={af.flatten()[rel_idx]:.4g}  liger={bf.flatten()[rel_idx]:.4g}"
+        )
+        assert torch.allclose(a, b, atol=atol, rtol=rtol), (
+            f"{label}: max_abs={diff.max():.4g}  max_rel={rel.max():.4g}  (atol={atol}, rtol={rtol})"
+        )
+
+    _assert_close(y1, y2, "forward output")
```
Comment (Tcc0403) [inline test/transformers/test_swiglu.py:297]: https://docs.pytorch.org/docs/2.11/testing.html#torch.testing.assert_close ditto

File: benchmark/scripts/benchmark_fused_moe.py [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+device = infer_device()
+
+
+# ---------------------------------------------------------------------------
+# HuggingFace reference: Python loop per expert
+# Matches transformers.models.mixtral.modeling_mixtral.MixtralSparseMoeBlock
+# ---------------------------------------------------------------------------
+
+
+def _huggingface_moe_forward(x, gate_up_proj, down_proj, top_k_index, top_k_weights):
```
Comment (Tcc0403) [inline benchmark/scripts/benchmark_fused_moe.py:40]: Import it from the test suite for maintainability, keeping it a single source of truth.

File: benchmark/scripts/benchmark_fused_moe.py [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+    "qwen3-moe-235b-ep32": MoEBenchConfig(T=128, E=4, H=7168, intermediate_dim=2560, K=8),
+    # DeepSeek-V3/R1 — EP=32 and EP=64
+    "deepseek-v3-ep32": MoEBenchConfig(T=128, E=8, H=7168, intermediate_dim=2048, K=8),
+    "deepseek-v3-ep64": MoEBenchConfig(T=64, E=4, H=7168, intermediate_dim=2048, K=8),
+}
+
+DEFAULT_MOE_MODEL = "qwen3-moe-30b"
+
+# Expert counts used in the num_experts sweep (independent of model).
+EXPERT_SWEEP_VALUES = [8, 16, 32, 64, 128]
```
Comment (Tcc0403) [inline benchmark/scripts/benchmark_fused_moe.py]: I wonder how we handle moe model configuration. Should it be a centric config as dense model configs as we currently have in https://github.com/linkedin/Liger-Kernel/blob/main/benchmark/scripts/benchmark_model_configs.py ? WDYT? cc @noemotiovon 

https://github.com/linkedin/Liger-Kernel/blob/main/benchmark/scripts/benchmark_model_configs.py

**Review by Tcc0403 — COMMENTED** [review]
I haven't looked closely at the kernel implementation. I only left some comments on code organization that we can easily fix.

File: benchmark/scripts/benchmark_fused_moe.py (hunk 7 above)
Comment (Mecoli1219) [reply benchmark/scripts/benchmark_fused_moe.py:40]: I preferred not to make `benchmark` depends on `test` just for MoE so that benchmark can run independently.

File: test/transformers/test_fused_moe.py (hunk 5 above)
Comment (Mecoli1219) [reply test/transformers/test_fused_moe.py]: We should follow the convention

File: benchmark/scripts/benchmark_fused_moe.py (hunk 8 above)
Comment (Mecoli1219) [reply benchmark/scripts/benchmark_fused_moe.py]: We should definitely do it. I'll create a new dataclass under `benchmark_model_configs.py`  for that.

**Comment by Mecoli1219** [comment]
Hi @Tcc0403. Thanks for your review. Updated the code and address the comments!

**Review by kolehma8 — DISMISSED** [review]

**Review by kolehma8 — APPROVED** [review]

**Comment by michaelroyzen** [comment]
Thank you for this commit @Mecoli1219 @Tcc0403. It is well-intentioned, however it broke training with Qwen3.5 MoE. Can you please add tests specifically for that model?

**Comment by Mecoli1219** [comment]
Hi @michaelroyzen. Thanks for reporting and sorry for the inconvenience. I'm investigating it.

**Comment by Mecoli1219** [comment]
@michaelroyzen What specific error did you get and how to reproduce it? We do have unit test for qwen3.5-moe here https://github.com/linkedin/Liger-Kernel/blob/1faa0131db3dd0b2fe96ee200a293e342ef29365/test/transformers/test_monkey_patch.py#L2987-L3021
and convergence test at https://github.com/linkedin/Liger-Kernel/blob/1faa0131db3dd0b2fe96ee200a293e342ef29365/test/convergence/bf16/test_mini_models.py#L2270-L2287

**Comment by michaelroyzen** [comment]
Thanks @Mecoli1219, will get back to you with the exact trace. What I do know so far is that the b7025a70279b7bea1e635a92f20173abb66498a0 commit works but the current one does not.

**Comment by Mecoli1219** [comment]
Thanks! If you don't mind, you can create a issue to report it, and we can discuss more over there!

**Comment by michaelroyzen** [comment]
Hi @Mecoli1219, apologies for the delay. I've prepared a more detailed reproduction and action plan for a fix in #1246
